### PR TITLE
Disallow moving/copying dossiers to a branch node.

### PR DESCRIPTION
**Pull request by phgross**
Check allowedContentTypes before pasting objects from clipboard. To avoid pasting dossier in to branch_nodes.

Besides the repository add form shows now a warning message when adding a repositoryfolder to a leafnode, which already contains dossiers.
bildschirmfoto 2016-10-18 um 12 26 39

@deiferni

### CODE REVIEW

File: docs/HISTORY.txt [hunk 1]
```diff
@@ -5,6 +5,14 @@ Changelog
 4.11.0 (unreleased)
 -------------------
 
+- Show warning message when adding a repositoryfolder to a leafnode,
```
Comment (deiferni) [inline docs/HISTORY.txt]: ~~,~~

File: docs/HISTORY.txt [hunk 2]
```diff
@@ -5,6 +5,14 @@ Changelog
...
+  which already contains dossiers.
+  [phgross]
+
+- Check allowedContentTypes before pasting objects from clipboard.
+  To avoid pasting dossier in to branch_nodes.
```
Comment (deiferni) [inline docs/HISTORY.txt]: clipboard~~. T~~to

Comment (deiferni) [inline docs/HISTORY.txt]: branch nodes

Comment (deiferni) [inline docs/HISTORY.txt]: into

File: opengever/base/browser/paste.py [hunk 3]
```diff
@@ -32,13 +32,31 @@ def render(self):
                                     request=self.request, type='error')
             return self.redirect()
 
+        if not self.are_contenttypes_allowed(objs):
+            msg = _(u"msg_pasting_not_allowed",
+                    default=u"Can't paste items, its not allowed to add "
```
Comment (deiferni) [inline opengever/base/browser/paste.py]: it's

http://www.its-not-its.info/ 😛

File: opengever/base/browser/paste.py [hunk 4]
```diff
@@ -32,13 +32,31 @@ def render(self):
...
         msg = _(u"msg_successfuly_pasted",
                 default=u"Objects from clipboard successfully pasted.")
         api.portal.show_message(message=msg, request=self.request, type='info')
         return self.redirect()
 
+    def are_contenttypes_allowed(self, objs):
+        allowed_content_types = [
+            fti.id for fti in self.context.allowedContentTypes()]
+
+        for obj in objs:
```
Comment (deiferni) [inline opengever/base/browser/paste.py]: you could use `all`

e.g. `all(obj.portal_type in allowed_content_types for obj in objs)`

File: opengever/base/browser/paste.py [hunk 5]
```diff
@@ -32,13 +32,31 @@ def render(self):
...
+            return self.redirect()
+
         self.activate_request_layer()
         self.copy_objects(objs)
         msg = _(u"msg_successfuly_pasted",
                 default=u"Objects from clipboard successfully pasted.")
         api.portal.show_message(message=msg, request=self.request, type='info')
         return self.redirect()
 
+    def are_contenttypes_allowed(self, objs):
```
Comment (deiferni) [inline opengever/base/browser/paste.py]: i would call this `is_pasting_allowed` or similar (there might already be a view with this name, not 100% sure though, beware of conflicts!)

Comment (deiferni) [inline opengever/base/browser/paste.py]: hm or maybe `is_pasting_objects_allowed`?

File: opengever/base/locales/de/LC_MESSAGES/opengever.base.po [hunk 6]
```diff
@@ -304,17 +304,22 @@ msgid "msg_empty_clipboard"
...
 #. Default: "No items available"
-#: ./opengever/base/widgets.py:49
+#: ./opengever/base/widgets.py:50
 msgid "msg_no_items_available"
 msgstr "Es sind keine Elemente verfügbar."
 
+#. Default: "Can't paste items, its not allowed to add objects of this type."
+#: ./opengever/base/browser/paste.py:36
+msgid "msg_pasting_not_allowed"
+msgstr "Es können keine Objekte eingefügt werden, es ist nicht erlaubt Objekte dieses Typs einzufügen."
```
Comment (deiferni) [inline opengever/base/locales/de/LC_MESSAGES/opengever.base.po]: this sounds a bit redundant, i'd only use

Es ist nicht erlaubt Objekte dieses Typs einzufügen.

File: opengever/base/tests/test_copy_paste.py [hunk 7]
```diff
@@ -159,3 +159,25 @@ def test_object_renaming_is_not_journalized(self, browser):
                           'Dossier modified: dossier-1',
                           u'Document added: Testdokum\xe4nt',
                           'Dossier added: dossier-1'], titles)
+
+    @browsing
+    def test_pasting_dossiers_into_a_branch_node_redirects_back_and_show_statusmessage(self, browser):
```
Comment (deiferni) [inline opengever/base/tests/test_copy_paste.py]: shows

File: opengever/base/tests/test_copy_paste.py [hunk 8]
```diff
@@ -159,3 +159,25 @@ def test_object_renaming_is_not_journalized(self, browser):
...
+        """When a object gets copied, it has to be renamed after the creation,
```
Comment (deiferni) [inline opengever/base/tests/test_copy_paste.py]: ✂️  copy-pasted comment?

File: opengever/repository/locales/de/LC_MESSAGES/opengever.repository.po [hunk 9]
```diff
@@ -144,6 +144,11 @@ msgstr "Status"
 msgid "msg_delete_confirmation"
 msgstr "Möchten Sie die aktuelle Ordnungsposition wirklich löschen?"
 
+#. Default: "You are adding a repositoryfolder to a leafnode which contains dossiers. This is only temporarily allowed and all dossiers must be moved in to the new leafnode."
+#: ./opengever/repository/repositoryfolder.py:194
+msgid "msg_leafnode_warning"
+msgstr "Sie fügen eine Ordnungsposition einem Blattknoten hinzu, der Dossiers enthält. Dies ist nur vorübergehend erlaubt und alle enthaltenden Dossiers müssen anschliessend in den neuen Blattknoten verschoben werden."
```
Comment (deiferni) [inline opengever/repository/locales/de/LC_MESSAGES/opengever.repository.po]: der bereits Dossiers enthält
tönt evtl etwas besser?

Comment (deiferni) [inline opengever/repository/locales/de/LC_MESSAGES/opengever.repository.po]: enthaltenden Dossiers müssen anschliessend in den neuen Blattknoten verschoben werden.
-> 
darin enthaltenen Dossiers müssen anschliessend wieder in einen Blattknoten verschoben werden.

File: opengever/repository/repositoryfolder.py [hunk 10]
```diff
@@ -187,6 +189,25 @@ def _filterer(fti):
 class AddForm(TranslatedTitleAddForm):
     grok.name('opengever.repository.repositoryfolder')
 
+    def render(self):
+        if self.contains_dossiers():
+            msg = _('msg_leafnode_warning',
+                    default=u'You are adding a repositoryfolder to a leafnode '
+                    'which contains dossiers. This is only temporarily '
```
Comment (deiferni) [inline opengever/repository/repositoryfolder.py]: You are adding a repositoryfolder to a leafnode which already contains dossiers. This is only temporarily  allowed and all dossiers must be moved into a new leafnode afterwards.

Comment (phgross) [inline opengever/repository/repositoryfolder.py]: 👍

**Comment by phgross** [comment]
@deiferni updated ...

**Pull request merged**